[Rando] Enemy Soul Shuffle

Caladius opened the pull request:
~~This forces you to obtain an Enemy Soul Item before the Enemy will spawn in the world, similar to Boss Souls.~~

All Soul'd enemies will be invincible by default, once you obtain their soul they will take damage as normal.
- This adds a new VB_APPLY_DAMAGE_TO_ACTOR (thanks Proxy) that is used to negate damage as needed. In the future this can also be used with Damage modifiers. (we have one currently that can be updated in a separate PR.

The goal is for this to go along with https://github.com/HarbourMasters/2ship2harkinian/pull/1282 but can be played alone if desired.

Enemies Remaining:
- [x] Like-Like (Highly Segmented)
- [x] Freezard
- [x] Red Bubble
- [x] Flying pot
- [x] Nejiron
- [x] Blue Bubble


Somethings to Note:
- [x] - Stalchild Eyes will disappear in some situations like having a Junk Item on the shelf under it in a shop (investigating).
- I thought I had more for here but guess not... Have fun!

<!--- section:artifacts:start -->
### Build Artifacts
  - [2ship-windows.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/5004284056.zip)
  - [2ship-linux.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/5004307204.zip)
  - [2ship-mac.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/5004560735.zip)
<!--- section:artifacts:end -->

Caladius committed:
Add Aliens and initial code moved from old build.

Caladius committed:
Update Enemy Souls to latest Develop and fix issues.

Caladius committed:
Change Stalchild Animation to Idle

Caladius committed:
Fix Stalchild Eyes

Caladius committed:
Adds Hiploop to keep in line with Enemy Drops

Caladius committed:
Update to latest Dev

Caladius committed:
correction

Eblo commented on mm/2s2h/Rando/ActorBehavior/Souls.cpp:
```diff
@@ -26,47 +74,82 @@ void ShouldActorDraw(Actor* actor, bool* should, RandoInf randoInf) {
+
+    // ShouldActorDraw & ShouldActorUpdate for Enemy Souls
+    COND_HOOK(ShouldActorDraw, shouldEnemyRegister, [](Actor* actor, bool* should) {
+        if (actor->category != ACTORCAT_ENEMY || actor->category == ACTORCAT_BOSS) {
```
If enemies and bosses are separate, is this condition needed?

Caladius replied:
Both ShouldActorDraw and ShouldActorUpdate hooks will fire off whenever an Actor triggers it, I was under the assumption since the hook didnt specify a type of actor that I should be ensuring if filters out Bosses?

Eblo replied:
The `COND_HOOK` already has `shouldEnemyRegister` as the condition. My thought was that checking for boss types in the function was superfluous, but after checking I see that some non-boss enemies use that type.

Eblo commented on mm/2s2h/Rando/DrawFuncs.cpp:
```diff
@@ -20,52 +20,1089 @@ s32 EnMinifrog_OverrideLimbDraw(PlayState* play, s32 limbIndex, Gfx** dList, Vec
+    if (limbIndex == FIRE_KEESE_LIMB_HEAD) {
+        Gfx* gfx = play->state.gfxCtx->polyXlu.p;
+        Scene_SetRenderModeXlu(play, 1, 2);
+        gSPMatrix(gfx++, Matrix_NewMtx(play->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
```
```suggestion
        MATRIX_FINALIZE_AND_LOAD(gfx++, play->state.gfxCtx);
```

Eblo commented on mm/2s2h/Rando/DrawFuncs.cpp:
```diff
@@ -20,52 +20,1089 @@ s32 EnMinifrog_OverrideLimbDraw(PlayState* play, s32 limbIndex, Gfx** dList, Vec
+    if (limbIndex == REAL_BOMBCHU_LIMB_TAIL_END) {
+        OPEN_DISPS(play->state.gfxCtx);
+        Matrix_ReplaceRotation(&play->billboardMtxF);
+        gSPMatrix(POLY_OPA_DISP++, Matrix_NewMtx(play->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
```
```suggestion
        MATRIX_FINALIZE_AND_LOAD(POLY_OPA_DISP++, play->state.gfxCtx);
```

Eblo commented on mm/2s2h/Rando/DrawFuncs.cpp:
```diff
@@ -20,52 +20,1089 @@ s32 EnMinifrog_OverrideLimbDraw(PlayState* play, s32 limbIndex, Gfx** dList, Vec
+        Matrix_RotateZYX(0x4000, 0, 0, MTXMODE_APPLY);
+        gDPSetEnvColor(POLY_OPA_DISP++, 0, 0, 80, 255);
+        gDPSetPrimColor(POLY_OPA_DISP++, 0, 0, 10, 0, 40, 255);
+        gSPMatrix(POLY_OPA_DISP++, Matrix_NewMtx(play->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
```
```suggestion
        MATRIX_FINALIZE_AND_LOAD(POLY_OPA_DISP++, play->state.gfxCtx);
```

Eblo commented on mm/2s2h/Rando/DrawFuncs.cpp:
```diff
@@ -20,52 +20,1089 @@ s32 EnMinifrog_OverrideLimbDraw(PlayState* play, s32 limbIndex, Gfx** dList, Vec
+    }
+
+    gSPDisplayList(&gfx[0], gSetupDLs[SETUPDL_25]);
+    gSPMatrix(&gfx[1], Matrix_NewMtx(gPlayState->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
```
```suggestion
    MATRIX_FINALIZE_AND_LOAD(&gfx[1], gPlayState->state.gfxCtx);
```

Eblo commented on mm/2s2h/Rando/DrawFuncs.cpp:
```diff
@@ -20,52 +20,1089 @@ s32 EnMinifrog_OverrideLimbDraw(PlayState* play, s32 limbIndex, Gfx** dList, Vec
+    gSPDisplayList(POLY_OPA_DISP++, gSetupDLs[SETUPDL_25]);
+    gDPSetPrimColor(POLY_OPA_DISP++, 0, 0xFF, 0, 0, 0, 255);
+    gSPSegment(POLY_OPA_DISP++, 0x08, (uintptr_t)D_801AEFA0);
+    gSPMatrix(POLY_OPA_DISP++, Matrix_NewMtx(gPlayState->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
```
```suggestion
   MATRIX_FINALIZE_AND_LOAD(POLY_OPA_DISP++, gPlayState->state.gfxCtx);
```

Eblo commented on mm/2s2h/Rando/DrawFuncs.cpp:
```diff
@@ -20,52 +20,1089 @@ s32 EnMinifrog_OverrideLimbDraw(PlayState* play, s32 limbIndex, Gfx** dList, Vec
+    gDPSetEnvColor(POLY_XLU_DISP++, 255, 255, 255, 255);
+    gSPDisplayList(POLY_XLU_DISP++, (Gfx*)gBlackBoeBodyMaterialDL);
+    Matrix_ReplaceRotation(&gPlayState->billboardMtxF);
+    gSPMatrix(POLY_XLU_DISP++, Matrix_NewMtx(gPlayState->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
```
```suggestion
    MATRIX_FINALIZE_AND_LOAD(POLY_XLU_DISP++, gPlayState->state.gfxCtx);
```

Eblo commented on mm/2s2h/Rando/DrawFuncs.cpp:
```diff
@@ -20,52 +20,1089 @@ s32 EnMinifrog_OverrideLimbDraw(PlayState* play, s32 limbIndex, Gfx** dList, Vec
+    gDPSetPrimColor(POLY_XLU_DISP++, 0, 0xFF, 245, 97, 0, 255);
+    gSPDisplayList(POLY_XLU_DISP++, (Gfx*)gBlackBoeEyesDL);
+    Matrix_Scale(0.009f, 0.009f, 0.009f, MTXMODE_APPLY);
+    gSPMatrix(POLY_XLU_DISP++, Matrix_NewMtx(gPlayState->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
```
```suggestion
    MATRIX_FINALIZE_AND_LOAD(POLY_XLU_DISP++, gPlayState->state.gfxCtx);
```

Eblo commented on mm/2s2h/Rando/DrawFuncs.cpp:
```diff
@@ -20,52 +20,1089 @@ s32 EnMinifrog_OverrideLimbDraw(PlayState* play, s32 limbIndex, Gfx** dList, Vec
 
-    MATRIX_FINALIZE_AND_LOAD(POLY_XLU_DISP++, gPlayState->state.gfxCtx);
-    gSPDisplayList(POLY_XLU_DISP++, sp68);
+    gSPMatrix(POLY_XLU_DISP++, Matrix_NewMtx(gPlayState->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
```
```suggestion
    MATRIX_FINALIZE_AND_LOAD(POLY_XLU_DISP++, gPlayState->state.gfxCtx);
```

Eblo commented on mm/2s2h/Rando/GiveItem.cpp:
```diff
@@ -271,7 +271,39 @@ void Rando::GiveItem(RandoItemId randoItemId) {
         case RI_SOUL_TWINMOLD:
-            Flags_SetRandoInf(RANDO_INF_OBTAINED_SOUL_OF_GOHT + (randoItemId - RI_SOUL_GOHT));
+            // Flags_SetRandoInf(RANDO_INF_OBTAINED_SOUL_OF_GOHT + (randoItemId - RI_SOUL_GOHT));
+            // break;
```
```suggestion
```

Eblo commented on mm/2s2h/Rando/ActorBehavior/Souls.cpp:
```diff
@@ -10,6 +12,52 @@ extern "C" {
+    { RI_SOUL_MAD_SCRUB,    { DrawMadScrub,     { ACTOR_EN_DEKUNUTS }, RANDO_INF_OBTAINED_SOUL_OF_MAD_SCRUBS } },
+    { RI_SOUL_OCTOROK,      { DrawOctorok,      { ACTOR_EN_OKUTA }, RANDO_INF_OBTAINED_SOUL_OF_OCTOROKS } },
+    { RI_SOUL_PEAHAT,       { DrawPeahat,       { ACTOR_EN_PEEHAT }, RANDO_INF_OBTAINED_SOUL_OF_PEAHATS } },
+    { RI_SOUL_REDEAD,       { DrawRedead,       { ACTOR_EN_RD, ACTOR_EN_RAILGIBUD }, RANDO_INF_OBTAINED_SOUL_OF_REDEADS } },
```
Should this include `ACTOR_EN_RAILGIBUD` (Gibdos patrolling Music Box House)?

Caladius replied:
What is the reason for not including it? (Open to removing it, just unsure why)

Eblo replied:
I was mentally separating Redeads and Gibdos, but it makes sense to keep them together. My new question is whether we should include `ACTOR_EN_TALK_GIBUD`, which is the trade Gibdos in the well. If we include them, that has logical implications for the well. If we leave them separate, then enemy drops may have to account for that. I'm leaning slightly towards including them and updating well logic, but I'm open to arguments for leaving them excluded.

Caladius replied:
I purposely left those out, not against including them but it just seemed like an entire region lock out for a soul was a lil much. Your call here.

Eblo replied:
Between this and the drops review I'm doing, let's keep 'em excluded.

Eblo commented on mm/2s2h/Rando/ActorBehavior/Souls.cpp:
```diff
@@ -10,6 +12,52 @@ extern "C" {
+    { RI_SOUL_SKULLTULA,    { DrawSkulltula,    { ACTOR_EN_ST, ACTOR_EN_SW }, RANDO_INF_OBTAINED_SOUL_OF_SKULLTULAS } },
+    { RI_SOUL_SLIME,        { DrawSlime,        { ACTOR_EN_SLIME }, RANDO_INF_OBTAINED_SOUL_OF_SLIMES } },
+    { RI_SOUL_SNAPPER,      { DrawSnapper,      { ACTOR_EN_KAME }, RANDO_INF_OBTAINED_SOUL_OF_SNAPPERS } },
+    { RI_SOUL_STALCHILD,    { DrawStalchild,    { ACTOR_EN_SKB, ACTOR_EN_RAIL_SKB }, RANDO_INF_OBTAINED_SOUL_OF_STALCHILDREN } },
```
Should this include `ACTOR_EN_HINT_SKB`?

Caladius replied:
Same question here as above.

Eblo replied:
Similar response as the above, though I don't think this is logically tangled as well Gibdos.

Caladius replied:
I guess they do lock out Beneath the Graveyard in the same way as the Gibdos would lock out Beneath the Well... No preference here, your call again haha.

Eblo replied:
Going with the above, let's leave this as-is.

Eblo replied:
No wait, you're right. The spider house ones can be ignored, but this graveyard case does matter. Let's remove it for now.

Eblo reviewed (changes requested):
Would love to test, but the builds broke because of references to `Matrix_NewMtx`. Left suggestions for cleanup as well as questions.

Caladius committed:
Merge remote-tracking branch 'origin/develop' into develop-enemy-souls-v2

Caladius committed:
Update Matrix Calls

Caladius committed:
Remove comment

Eblo commented on mm/2s2h/Rando/Menu.cpp:
```diff
@@ -323,8 +323,8 @@ static void DrawItemsTab() {
-    CVarCheckbox("Enemy Souls", "gPlaceholderBool",
-                 CheckboxOptions({ { .disabled = true, .disabledTooltip = "Coming Soon" } }));
+    CVarCheckbox("Enemy Souls", Rando::StaticData::Options[RO_SHUFFLE_ENEMY_SOULS].cvar,
+                 CheckboxOptions({ { .disabled = false, .disabledTooltip = "Coming Soon" } }));
```
This should be marked as incompatible with vanilla and French vanilla. Clang format might hate my suggestion:

```suggestion
                 CheckboxOptions({ { .tooltip = "Adds the \"souls\" of regular enemies to the item pool. Enemy Souls are items "
                            "that must be found in order for their corresponding enemy to spawn.",
                            .disabled = IncompatibleWithLogicSetting(RO_SHUFFLE_ENEMY_SOULS),
                            .disabledTooltip = "Incompatible with current Logic Setting" } }));
```

Eblo commented on mm/2s2h/Rando/ActorBehavior/Souls.cpp:
```diff
@@ -10,6 +12,52 @@ extern "C" {
+    { RI_SOUL_IRON_KNUCKLE, { DrawIronKnuckle,  { ACTOR_EN_IK }, RANDO_INF_OBTAINED_SOUL_OF_IRON_KNUCKLES } },
+    { RI_SOUL_KEESE,        { DrawKeese,        { ACTOR_EN_FIREFLY }, RANDO_INF_OBTAINED_SOUL_OF_KEESE } },
+    { RI_SOUL_LEEVER,       { DrawLeever,       { ACTOR_EN_NEO_REEBA }, RANDO_INF_OBTAINED_SOUL_OF_LEEVERS } },
+    { RI_SOUL_MAD_SCRUB,    { DrawMadScrub,     { ACTOR_EN_DEKUNUTS }, RANDO_INF_OBTAINED_SOUL_OF_MAD_SCRUBS } },
```
Missing the following enemies (possibly more than this):

- Like-Like
- Freezard
- Red Bubble
- Flying pot
- Nejiron
- Eyegore
- Dexihand
- Blue Bubble
- Poe
- Takkuri

Caladius replied:
Updated description to include these missing enemies, if anyone wants to contribute Souls please do.

Eblo commented on mm/2s2h/Rando/ActorBehavior/Souls.cpp:
```diff
@@ -10,6 +12,52 @@ extern "C" {
+    { RI_SOUL_KEESE,        { DrawKeese,        { ACTOR_EN_FIREFLY }, RANDO_INF_OBTAINED_SOUL_OF_KEESE } },
+    { RI_SOUL_LEEVER,       { DrawLeever,       { ACTOR_EN_NEO_REEBA }, RANDO_INF_OBTAINED_SOUL_OF_LEEVERS } },
+    { RI_SOUL_MAD_SCRUB,    { DrawMadScrub,     { ACTOR_EN_DEKUNUTS }, RANDO_INF_OBTAINED_SOUL_OF_MAD_SCRUBS } },
+    { RI_SOUL_OCTOROK,      { DrawOctorok,      { ACTOR_EN_OKUTA }, RANDO_INF_OBTAINED_SOUL_OF_OCTOROKS } },
```
This affects the freezable Octoroks in Ikana and GBT. Need to logically account for that.

Eblo commented on mm/2s2h/Rando/ActorBehavior/Souls.cpp:
```diff
@@ -10,6 +12,52 @@ extern "C" {
+    { RI_SOUL_DINOLFOS,     { DrawDinolfos,     { ACTOR_EN_DINOFOS }, RANDO_INF_OBTAINED_SOUL_OF_DINOLFOS } },
+    { RI_SOUL_DODONGO,      { DrawDodongo,      { ACTOR_EN_DODONGO }, RANDO_INF_OBTAINED_SOUL_OF_DODONGOS } },
+    { RI_SOUL_EENO,         { DrawEeno,         { ACTOR_EN_SNOWMAN }, RANDO_INF_OBTAINED_SOUL_OF_EENOS } },
+    { RI_SOUL_GARO,         { DrawGaroMaster,   { ACTOR_EN_JSO2, ACTOR_EN_JSO }, RANDO_INF_OBTAINED_SOUL_OF_GARO_MASTERS } },
```
This breaks regular Garos. The spawn cutscene never happens, but an invisible fire wall spawns.

Caladius replied:
Removed for now

Eblo reviewed (changes requested):
Thoughts:

- Some enemies are excluded. It almost seems like mini-bosses are intentional, but some of those are included. What is the reasoning? I imagine getting locked in a room where the only way out is to defeat all enemies is part of it, but that's functionally no different from how boss souls already work.
-  `CanKillEnemy` must return false if the requested enemy's soul is not acquired, just like boss souls. Maybe this can be simplified with an early return, using `soulMap` in some way.
- Places where enemies are a nuisance need not be defeated, such as the Bad Bat Tree of Doom, are more permissive if the relevant souls are not acquired. I don't think this is a problem, just worth noting.

Caladius committed:
Remove Garo for now

Caladius committed:
Add Octorok Soul req for Upper Canyon Access

Caladius committed:
Add soul requirement for Red Switch

Caladius committed:
Grant Soul INF when the option is off

Caladius committed:
Adds Nejiron Soul

Caladius committed:
Add Flying Pot Soul

mckinlee commented on mm/2s2h/Rando/DrawFuncs.cpp:
```diff
@@ -20,52 +20,1102 @@ s32 EnMinifrog_OverrideLimbDraw(PlayState* play, s32 limbIndex, Gfx** dList, Vec
+/* Keese */         #include "assets/objects/object_firefly/object_firefly.h"
+/* Leever */        #include "assets/objects/object_rb/object_rb.h"
+/* Mad Scrub */     #include "assets/objects/object_dekunuts/object_dekunuts.h"
+/* Nejiron */       #include "assets//objects/object_gmo/object_gmo.h"
```
Is the extra slash needed?

mckinlee commented on mm/2s2h/Rando/MiscBehavior/OnFileCreate.cpp:
```diff
@@ -399,6 +399,13 @@ void Rando::MiscBehavior::OnFileCreate(s16 fileNum) {
 
+                // Give INF for Enemy Soul if the option is OFF
+                if (RANDO_SAVE_OPTIONS[RO_SHUFFLE_ENEMY_SOULS] == RO_GENERIC_NO) {
+                    for (int i = RANDO_INF_OBTAINED_SOUL_OF_ARMOS; i < RANDO_INF_OBTAINED_SOUL_OF_WOLFOS; i++) {
```
Should this be `<=` otherwise `RANDO_INF_OBTAINED_SOUL_OF_WOLFOS` would get skipped?

mckinlee commented on mm/2s2h/Rando/ActorBehavior/Souls.h:
```diff
@@ -0,0 +1,8 @@
+
+extern std::unordered_map<RandoItemId, std::tuple<std::function<void()>, std::vector<ActorId>, RandoInf>> soulMap;
+
+#endif // SOULS_
```
Shouldn't this be `SOULS_H`?

mckinlee reviewed (commented):
Since I was already in the code, I made note of a few things. Nothing crazy. That aside, this is really cool and I can't wait to see this partnered with the other shuffles!

mckinlee committed:
ew (#14)

Caladius committed:
cleanup

Caladius committed:
Merge branch 'develop-enemy-souls-v2' of https://github.com/Caladius/2ship2harkinian into develop-enemy-souls-v2

Caladius committed:
Update to latest Dev

Caladius committed:
Add Bubble Soul

Caladius committed:
Adds Freezard

Caladius committed:
Change Soul behaivor to Invincible

Eblo commented on mm/2s2h/Rando/ActorBehavior/Souls.cpp:
```diff
@@ -1,5 +1,8 @@
 #include "ActorBehavior.h"
 #include <libultraship/bridge/consolevariablebridge.h>
+#include "Rando/DrawFuncs.h"
+#include "spdlog/spdlog.h"
```
```suggestion
```

Eblo commented on mm/src/code/z_collision_check.c:
```diff
@@ -3526,12 +3526,15 @@ void CollisionCheck_ApplyDamage(struct PlayState* play, CollisionCheckContext* c
-            if (col->actor->colChkInfo.damage < finalDamage) {
-                col->actor->colChkInfo.damage = finalDamage;
+        if (GameInteractor_Should(VB_APPLY_DAMAGE_TO_ACTOR, true, col->actor, col->actor->colChkInfo.damageEffect,
+                                  col->actor->colChkInfo.damage, elem->acDmgInfo.dmgFlags)) {
```
Since you pass in the actor, you don't need to pass in the actor's colChkInfo fields too. You could just extrapolate those in your enhancement code.

Eblo reviewed (commented):
I'm not sure what, but something about the way this was done causes it to only work if boss souls are also shuffled. If one tries to only shuffle enemy souls, glitchless seeds will fail to generate, and no logic seeds will generate but with no souls actually in the pool.

I haven't exhaustively tested everything, but it seems like Eeno souls simply are not in my seed. Eeno drops contained a major item, but there was no way for me to defeat one to get it.

On the flip side, aside from Real Bombchus, I didn't happen upon any enemies that I could defeat without having their souls first. Octoroks and ChuChus even shrugged off ice arrows.

Caladius committed:
header cleanup and arg consolidation

mckinlee commented:
> I'm not sure what, but something about the way this was done causes it to only work if boss souls are also shuffled. If one tries to only shuffle enemy souls, glitchless seeds will fail to generate, and no logic seeds will generate but with no souls actually in the pool.
> 
> I haven't exhaustively tested everything, but it seems like Eeno souls simply are not in my seed. Eeno drops contained a major item, but there was no way for me to defeat one to get it.
> 
> On the flip side, aside from Real Bombchus, I didn't happen upon any enemies that I could defeat without having their souls first. Octoroks and ChuChus even shrugged off ice arrows.

It doesn't look like the enemy soul items are added to the item pool, so the generator can't place them anywhere. Logic still expects those flags, so "enemy souls only" fails since the pool lacks any soul items, yet damage is gated on the corresponding INF flag explains why actors seem to be near invulnerable. I think we just need to mirror the boss souls block around lines 241 to 249 of OnFileCreate.cpp, maybe?

Caladius committed:
Fixes OnFileCreate to actually shuffle Enemy Souls. took this op. to fix up Boss Souls shuffling as it was needed to cleanly shuffle Enemy Souls.

Eblo commented on mm/2s2h/Rando/StaticData/Items.cpp:
```diff
@@ -168,11 +168,45 @@ std::map<RandoItemId, RandoStaticItem> Items = {
+    RI(RI_SOUL_SHELLBLADE,            "the",  "Soul of Shellblades",        RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_SOUL_SKULLFISH,             "the",  "Soul of Skullfish",          RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_SOUL_SKULLTULA,             "the",  "Soul of Skulltulas",         RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_SOUL_SLIME,                 "the",  "Soul of Slimes",             RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
```
I prefer the names match the in-game names
```suggestion
    RI(RI_SOUL_CHUCHU,                 "the",  "Soul of Chuchus",             RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
```

Eblo commented on mm/2s2h/Rando/StaticData/Items.cpp:
```diff
@@ -168,11 +168,45 @@ std::map<RandoItemId, RandoStaticItem> Items = {
+    RI(RI_SOUL_EENO,                  "the",  "Soul of Eenos",              RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_SOUL_FLYING_POT,            "the",  "Soul of Flying Pots",        RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_SOUL_FREEZARD,              "the",  "Soul of Freezards",          RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_SOUL_GRASSHOPPER,           "the",  "Soul of Grasshoppers",       RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
```
```suggestion
    RI(RI_SOUL_DRAGONFLY,           "the",  "Soul of Dragonflies",       RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
```

Eblo commented on mm/2s2h/Rando/StaticData/Items.cpp:
```diff
@@ -168,11 +168,45 @@ std::map<RandoItemId, RandoStaticItem> Items = {
+    RI(RI_SOUL_BAT,                   "the",  "Soul of Bats",               RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_SOUL_BEAMOS,                "the",  "Soul of Beamos",             RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_SOUL_BOE,                   "the",  "Soul of Boes",               RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_SOUL_BOMBCHU,               "the",  "Soul of Bombchus",           RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
```
In case we ever make souls for the item, for some reason:
```suggestion
    RI(RI_SOUL_REAL_BOMBCHU,               "the",  "Soul of Real Bombchus",           RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
```

Eblo commented on mm/2s2h/Rando/StaticData/Items.cpp:
```diff
@@ -168,11 +168,45 @@ std::map<RandoItemId, RandoStaticItem> Items = {
-    RI(RI_SOUL_ODOLWA,                "the",  "Soul of Odolwa",             RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
-    RI(RI_SOUL_TWINMOLD,              "the",  "Soul of Twinmold",           RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_SOUL_ARMOS,                 "the",  "Soul of Armos",              RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_SOUL_BAT,                   "the",  "Soul of Bats",               RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
```
```suggestion
    RI(RI_SOUL_BAD_BAT,                   "the",  "Soul of Bad Bats",               RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
```

Eblo commented on mm/2s2h/Rando/StaticData/Items.cpp:
```diff
@@ -168,11 +168,45 @@ std::map<RandoItemId, RandoStaticItem> Items = {
+    RI(RI_SOUL_STALCHILD,             "the",  "Soul of Stalchildren",       RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_SOUL_TEKTITE,               "the",  "Soul of Tektites",           RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+	RI(RI_SOUL_TWINMOLD,              "the",  "Soul of Twinmold",           RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_SOUL_WALLMASTER,            "the",  "Soul of Wallmasters",        RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
```
Since this applies to both Wallmasters and Floormasters, is there any sort of name to capture that instead of implying Wallmasters only? There's only one Floormaster in the game, so I don't think it's worth splitting out another soul.

Caladius replied:
Soul of Hands? I dont really know how to unify them lol

Eblo commented on mm/2s2h/Rando/DrawFuncs.cpp:
```diff
@@ -20,52 +20,1236 @@ s32 EnMinifrog_OverrideLimbDraw(PlayState* play, s32 limbIndex, Gfx** dList, Vec
+    DrawEnLight({ 155, 155, 155 }, { 10.0f, 10.0f, 10.0f });
+}
+
+extern void DrawIronKnuckle() {
```
Something about this is not right, as the Iron Knuckle soul does not render the mask holes.

Caladius replied:
Fixed this, was missing a DL.

Eblo commented on mm/src/code/z_collision_check.c:
```diff
@@ -3526,12 +3526,14 @@ void CollisionCheck_ApplyDamage(struct PlayState* play, CollisionCheckContext* c
-        if (!(col->acFlags & AC_HARD) || ((col->acFlags & AC_HARD) && (atElem->atDmgInfo.dmgFlags == 0x20000000))) {
-            if (col->actor->colChkInfo.damage < finalDamage) {
-                col->actor->colChkInfo.damage = finalDamage;
+        if (GameInteractor_Should(VB_APPLY_DAMAGE_TO_ACTOR, true, col->actor, elem->acDmgInfo.dmgFlags)) {
```
While this seems to work for the most part, it does not capture cases where enemies die from means other than damage. For example, Real Bomchus and flying pots are destroyed once they bump into anything, damaging or not. The absence of souls does not prevent that.

Bad Bats are a particular weird case where, while their death is prevented, the dying animation still happens. This results in a funny but confusing scenario where they spin in place endlessly without setting clear flags.

Eblo reviewed (commented):
Performed some cursory testing and review. This is far from exhaustive, and the main reason for that is that incoming enemy drops work will impact the direction on this.

Ideally we have some visual distinction to know that an enemy is in a soulless state, but I realize finding some one-size-fits-all adjustment is easier said than done. A translucent effect would be nice, but difficult. A lazy answer would be to draw sparkles after rendering the enemy normally, like we do to distinguish progressive magic from refills, or literally nothing.

Caladius committed:
Update to latest Dev

Caladius committed:
Naming Updates

Caladius committed:
Name Updates pt2

Caladius committed:
Add Tracker Icons and adds a Default to avoid erroring

Caladius committed:
Fix Iron Knuckle Helmet

Caladius committed:
revert item icon default and add soul flame to tatl target for enemies that are invincible.

Eblo committed:
Move VBs to GameInteractor_VanillaBehavior.h

Eblo committed:
Use VB_PERFORM_AC_COLLISION for enemy souls

Eblo committed:
Add enemy souls for closer parity with drops

Stub out enemy draw funcs
Use macros to streamline existing skeleton inits
Logically account for enemy souls in CanKillEnemy
Fix logical quirk with Ikana ice arrows

Eblo committed:
Draw everything

Eblo committed:
Remove unused INIT arg from SETUP_FLEX_SKEL

Eblo committed:
Add Igos, Keeta, Gomess

Make soul RIs contiguous, separate boss and enemy
Ignore target color for EnInvadepoh
Remove Real Bombchus and Flying Pots for now
Require Gekko and Snapper souls for beating both

Eblo committed:
Add Enemy Souls to item tracker

Eblo committed:
Merge remote-tracking branch 'upstream/develop' into enemy-souls-v2-finish

Caladius committed:
Merge pull request #17 from Eblo/enemy-souls-v2-finish

Finish off enemy souls

Eblo committed:
Merge remote-tracking branch 'upstream/develop' into enemy-souls-rebase

Eblo committed:
Merge remote-tracking branch 'upstream/develop' into enemy-souls-rebase

Eblo committed:
clang format

Caladius committed:
Merge pull request #18 from Eblo/enemy-souls-rebase

Merge develop into enemy souls
